File: leet/longest-palindromic-substring.ts
/**
 * https://leetcode.com/problems/longest-palindromic-substring/
 * Given a string s, return the longest palindromic substring in s.
 */
export function longestPalindrome(s: string): string {
  let result = "";
  for (let i = 0; i < s.length - result.length; i++) {
    for (let j = i; j < s.length; j++) {
      let start = i;
      let end = j;
      let isPalindrome = true;

      while (start < end) {
        if (s[start] !== s[end]) {
          isPalindrome = false;
          break;
        }
        start++;
        end--;
      }

      if (isPalindrome) {
        if (j - i + 1 > result.length) {
          result = s.substring(i, j + 1);
        }
      }
    }
  }
  return result;
}

export function isPalindrome(s: string) {
  let start = 0;
  let end = s.length - 1;

  while (start < end) {
    if (s[start] !== s[end]) return false;
    start++;
    end--;
  }

  return true;
}

File: returned/python/longest_palindromic_substring.py
# https://leetcode.com/problems/longest-palindromic-substring/description/

class Solution:
    def longestPalindrome(self, s: str) -> str:
        longest = ''

        for i in range(len(s)):
            # odd length
            l, r = i, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if(r-l+1 > len(longest)):
                    longest = s[l:r+1]
                l -= 1
                r += 1

            # even length
            l, r = i, i+1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if(r-l+1 > len(longest)):
                    longest = s[l:r+1]
                l -= 1
                r += 1

        return longest

